test/test.ads.js

```diff
@@ -889,85 +684,115 @@ QUnit.test('startLinearAdMode should only trigger adstart from correct states',
   var adstart = sinon.spy();
   this.player.on('adstart', adstart);
 
-  this.player.ads.state = 'preroll?';
   this.player.ads.startLinearAdMode();
-  assert.strictEqual(adstart.callCount, 1, 'preroll? state');
+  assert.strictEqual(adstart.callCount, 0, 'Before play');
+
+  this.player.trigger('play');
 
-  this.player.ads.state = 'content-playback';
   this.player.ads.startLinearAdMode();
-  assert.strictEqual(adstart.callCount, 2, 'content-playback state');
+  assert.strictEqual(adstart.callCount, 0, 'Before adsready');
 
-  this.player.ads.state = 'postroll?';
+  this.player.trigger('adsready');
   this.player.ads.startLinearAdMode();
-  assert.strictEqual(adstart.callCount, 3, 'postroll? state');
+  assert.strictEqual(adstart.callCount, 1, 'Preroll');
 
-  this.player.ads.state = 'content-set';
   this.player.ads.startLinearAdMode();
-  this.player.ads.state = 'ads-ready?';
+  assert.strictEqual(adstart.callCount, 1, 'During preroll playback');
+
+  this.player.ads.endLinearAdMode();
+  this.player.trigger('playing');
+
   this.player.ads.startLinearAdMode();
-  this.player.ads.state = 'ads-ready';
+  assert.strictEqual(adstart.callCount, 2, 'Midroll');
+
   this.player.ads.startLinearAdMode();
-  this.player.ads.state = 'ad-playback';
+  assert.strictEqual(adstart.callCount, 2, 'During midroll playback');
+
+  this.player.ads.endLinearAdMode();
+  this.player.trigger('playing');
+
+  this.player.trigger('ended');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(adstart.callCount, 3, 'Postroll');
+
   this.player.ads.startLinearAdMode();
-  assert.strictEqual(adstart.callCount, 3, 'other states');
+  assert.strictEqual(adstart.callCount, 3, 'During postroll playback');
+
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(adstart.callCount, 3, 'Ads done');
+  
 });
 
 QUnit.test('ad impl can notify contrib-ads there is no preroll', function(assert) {
 
-  this.player.ads.state = 'preroll?';
+  this.player.trigger('loadstart');
+  this.player.trigger('nopreroll');
```

Comment:
what happens if nopreroll is triggered after the first of `play` or `adsready`? i.e loadstart -> play -> nopreroll

Reply:
Added two more tests to cover these variants.